Drone: Add drone.bat and {before,after}-install.*

Flamefire opened the pull request:
This makes the default build work on Windows and adds customization points for work to be done before&after bootstrapping.

Also update the documentation accordingly.

See e.g. https://drone.cpp.al/boostorg/boost-ci/361/58/2 for a build using this.

sdarwin commented:
- In the refactored .drone.star there is the line `buildscript='boost'`  This causes a proliferation of file names to keep track of, since there is the previous `drone.sh` and now a new `boost.sh`.  It would be simpler to set `buildscript='drone'` the way it was before.   Consider that .drone.star and .drone.sh would be one file if they could be.  The only reason they are two separates files is because starlark is python and that format doesn't conveniently support bash and so it had to be split into two files.   

- boost.bat:  ideally, change name to drone.bat.   And should this include the commands to clone boost-ci, the way that drone.sh does?  See https://github.com/boostorg/json/blob/develop/.drone/drone.bat as an example.  Or has the new refactor made that step obsolete since you are downloading that somewhere else?    Add new windows drone tests to https://drone.cpp.al/boostorg/boost-ci .

- BOOST_CUSTOM_INSTALL_SCRIPT:    The way this is being run appears to be just to plainly "call" it `"$BOOST_CUSTOM_INSTALL_SCRIPT"`.    does that work?   Probably instead it should be called with the `source` command? Or the dot (.) command which is the same as `source`.

Flamefire commented:
>  It would be simpler to set buildscript='drone' the way it was before.

buildtype/buildscript set to 'boost' has always been the default since `functions.star` was added:
https://github.com/boostorg/boost-ci/commit/72c587dee78b5c53c464ed39244097d5cd017cd3#diff-d74bd46c0f0eab290139a5a4f9de7ac2391242af9842ffd76756efa072247c7dR11

And looks like we made a mistake that lead to those 2 files: In https://github.com/boostorg/boost-ci/pull/186 one initial objection was that `drone.sh` should not always source `common_install` while I argued that having the "usual" case of a "regular boost build" as easy as possible. This has led to me adding `boost.sh` with https://github.com/boostorg/boost-ci/commit/8e4680544d30e56859d4abe98990f23892557612 due to https://github.com/boostorg/boost-ci/blob/2a2ed9df5962ca98b66971d6d79194bbe68d3420/ci/drone/functions.star#L52 setting `buildtype="boost", buildscript=""` which will then call a `boost.sh` (and always has, see above) so it made sense for the default case to do the "default" build. But the PR has been forgotten and then merged after the commit so `boost.sh` and `drone.sh` are now the same. 

I guess I can add the changes to `drone.sh` and either delete `boost.sh` or make it default `$DRONE_JOB_BUILDTYPE=boost` and call drone.sh but I'd rather remove it. Or revert #186 and keep both files and the explicit `buildscript='boost'`. What do you think?

> boost.bat: ideally, change name to drone.bat.

This depends on the naming resolution above. If we choose to keep only a `drone.sh` then yes, I'll rename otherwise match `boost.sh` also in naming.

> And should this include the commands to clone boost-ci, the way that drone.sh does? See https://github.com/boostorg/json/blob/develop/.drone/drone.bat as an example. Or has the new refactor made that step obsolete since you are downloading that somewhere else?

That was always (?) superflous as it is done in `functions.star` already: https://github.com/boostorg/boost-ci/blob/2a2ed9df5962ca98b66971d6d79194bbe68d3420/ci/drone/functions.star#L153-L154
I find that a bit strange as it uses a different directory than e.g. GHA and doesn't have the same possibilities as the Unix-cases so the Boost-CI `windows-cxx-install.bat` is always used and you must have a` .drone/*.bat` build script in each repo as opposed to using the existing install-script if any and downloading the default build script from Boost-CI if no custom one exists.   
Hence that could/should be changed? I doubt anyone is using stuff from `.drone/boost-ci` so it should be possible to do that without breaking anything. Afterwards we could do the same cloning as in the `drone.sh`

However I actually do like the "cloning" of Boost-CI in `functions.star` as this allows a repo-specific build script to use the Boost-CI scripts w/o duplicating the clone step, especially the fail-safes/retry. The `custom_install` scripts serve a similar purpose: E.g. I need to generate a few locales or download ICU on Windows for Boost.Locale which is the only difference to a "vanilla" build. Hence those "plugins" allow to reuse everything else. So with that we may not need cloning in functions.star

> Add new windows drone tests to https://drone.cpp.al/boostorg/boost-ci 

I already did: The (above link)[https://drone.cpp.al/boostorg/boost-ci/356/58/2] is a Windows drone build using this new stuff. I factored this out so we can discuss those changes separate to the  `.drone.star` changes. Once we have this PR in

sdarwin commented:
> That was always (?) superflous as it is done in functions.star already

No, it was not superfluous because it was designed with separation of concerns.
functions.star was "installing packages" which travis used to do. In order to install packages, it needed to use some script. The scripts happen to be stored in boost-ci. But all it really needs is one file which is linux-cxx-install.sh or windows-cxx-install.sh. That could be cleaned up a little bit, but only downloading the one required file instead of the whole repo. We changed it for Linux. A similar change could be done on Windows and probably should be done.

Later on, boost-ci is download by the user-managed script to run the user-managed CI. But that is a distinct part of the process.

-----

> buildtype/buildscript set to 'boost' has always been the default since functions.star was added

No, you are mixing up buildtype and buildscript.

"boost" is very generic word and it could mean various things in these scripts, but what is has been set to mean is the standard "buildtype", which means when you build the gcc and clang tests. Not coverage. Not docs. 

If the buildtype is boost, it is running the standard gcc and clang tests only.

In the absence of a buildscript variable, in the early iterations of drone, it would load a script named boost, and inside of the script named boost it would only run the buildtype of boost. which means the gcc and clang tests.

If consolidating multiple buildtypes such as docs, and coverage,  etc into one script, that new script should have a name which doesn't overlap or conflict with the buildtype. Thus, it's called drone.sh.

sdarwin commented:
To throw another monkey wrench into the works, 
Last week, Alan de Freitas refactored drone yet again from a different angle.   
In certain ways, his changes are more extreme. 
It doesn't interfere with this pull request.
You could view it in the boostorg/url repository.  Maybe his new functions will go into a separate file `generate.star`, and offer yet another option.  

> However I actually do like the "cloning" of Boost-CI in functions.star 

consider something like generate.star instead of that.  

Actually, those functions are aimed at Vinnie's CPPAlliance repos, maybe too specifically.   A task could be to take that idea, and generalize it a bit, such that it could be used by any boost repo.

Flamefire commented:
> No, it was not superfluous because it was designed with separation of concerns.

Then it was the wrong word I meant that there was duplicated work (download of the repo twice) for no apparent reason. I now understand the motivation and it makes sense. I agree we should be using the same logic for Windows as on Linux (download a single file [or 2] if not existing yet).

> No, you are mixing up buildtype and buildscript.
> ...
> In the absence of a buildscript variable, in the early iterations of drone, it would load a script named boost, and inside of the script named boost it would only run the buildtype of boost. which means the gcc and clang tests.

And even when the buildscript variable [was introduced](https://github.com/boostorg/boost-ci/commit/d3ab40f67e93d36f66c199650a542373723b3624) it defaulted to the buildtype which in turn defaulted to "boost"

I still see that reasonable as coverage, sanitizer etc. builds all do a more or less default `b2 lib/test` build. And the script lives in the `.drone` folder, so having `.drone/boost.sh` with some buildtypes that all do a "Boost build" of some sort, even when collecting coverage while doing so sounds useful because now library maintainers can add a custom script for doing other things (like doc builds) in a separate and reasonable named file, e.g. `drone.sh` reusing the build type for different things.   
But this is also possible if the script is named `.drone/drone.sh` so I don't really mind that although I find `.drone/boost.sh` more descriptive than the former.

sdarwin commented:
> And even when the buildscript variable [was introduced](https://github.com/boostorg/boost-ci/commit/d3ab40f67e93d36f66c199650a542373723b3624) it defaulted to the buildtype which in turn defaulted to "boost"

In that scenario, if the buildtype was boost and thus the script was boost, that script would only contain the standard gcc and clang tests only, not coverage, not docs from other travis jobs.

>  find .drone/boost.sh more descriptive than the former.

If creating the system completely from scratch, it might be slightly better to name it .drone/boost.sh

But that's not the situation now since it's already been deployed.   `boost.sh` has a certain semantic meaning already.  It indicates the buildtype is boost, etc.     It is easier to keep the names the same as before, and the documentation, and not having to keep track of multiple versions.      `drone.sh` has been deployed for 2 years to multiple repos, it's not so bad, and it's easier to keep naming conventions the same.

Flamefire commented:
> But that's not the situation now since it's already been deployed. `boost.sh` has a certain semantic meaning already. It indicates the buildtype is boost, etc. It is easier to keep the names the same as before, and the documentation, and not having to keep track of multiple versions. `drone.sh` has been deployed for 2 years to multiple repos, it's not so bad, and it's easier to keep naming conventions the same.

But the situation here is that we have 2 files doing pretty much the same. Given your reasoning I'd propose stripping down `boost.sh` to the bare minimum with the assumption `DRONE_JOB_BUILDTYPE=boost`, i.e. remove all other actions (like coverage) and adding the changes from this PR (also) to `drone.sh`. Then rename the new boost.bat to drone.bat for consistency. Then in the [new drone.star branch](https://github.com/boostorg/boost-ci/blob/drone-refactor/.drone.star)  the default can be "`buildscript='drone', buildtype=None/boost`"

sdarwin commented:
Yes that sounds good.   Add the changes from this PR to drone.sh.   Maybe the file `boost.sh` can be removed, since new boost repositories, copying all the examples from boost-ci, would not use it.

Flamefire commented:
Checking the repos I found that some do use `.drone/drone.sh` files together with `before-install` and `before-script` files. So I renamed `custom_install` to `after-install` and added support for `before-install` (optional as well) and also source the files instead of executing them. This has the disadvantage that it might change the current working directory but that could be an advantage too. 

Names are chosen with this minor derivation because I think "before-install common-install after-install" makes more sense especially as on Windows "Script" is simply called "Compile" and I also checked that no-one is really using "before-script" (all empty) so it doesn't change anything. However I kept the existing "before-install" but "common_install" difference

Also changed the default to `buildscript=drone` and test with this PR is running at https://drone.cpp.al/boostorg/boost-ci/361

Flamefire commented:
@sdarwin Can you review the updated changeset and merge please?

sdarwin commented:
`drone.bat` should clone boost-ci, analogous to what drone.sh is doing.

sdarwin commented:
Thanks for the update to drone.bat. 

Next, in the latest change the tests are being run from $BOOST_CI_SRC_FOLDER instead of $BOOST_ROOT.  What had often been done previously was a $BOOST_ROOT is set up and configured, the library is copied into place in $BOOST_ROOT, and tests are run from that location.    

I guess what this change is implying, is that once a $BOOST_ROOT variable has been set, it doesn't really matter anymore from where you run the tests? 

Would it nevertheless be more clear to demonstrate where tests are running, for it to be in $BOOST_ROOT?  ..or it doesn't matter.

Flamefire commented:
> I guess what this change is implying, is that once a $BOOST_ROOT variable has been set, it doesn't really matter anymore from where you run the tests?

I basically just wanted to reference scripts in a uniform way. What we do in `common_install.*` is `cp -r $BOOST_CI_SRC_FOLDER/* $BOOST_ROOT/libs/$SELF`
In the bat-file we even go back to the library folder in BOOST_ROOT: https://github.com/boostorg/boost-ci/blob/62e542cbae8cc0a696382205f15876bfa635ebc2/ci/common_install.bat#L65-L66

Anyway we have the library at `$BOOST_CI_SRC_FOLDER` AND at `$BOOST_ROOT/libs/$SELF` so we could use either and the Boost.CI scripts always use `$BOOST_ROOT` so it indeed doesn't matter where the tests are run from as we have all that in the environment.

However thinking more about it it might make sense to always use  `$BOOST_ROOT` and pretend `$BOOST_CI_SRC_FOLDER` doesn't exist anymore after `common_install` so we can move to moving the library folder instead of copying it but I'm not sure if that might break other things like coverage. Especially as IIRC the glob we use to copy the folder contents doesn't include the hidden files/folders, especially `.git` which for the copy might be an advantage (faster).

But if we look at `drone.*` as it is now it makes more sense as it is (using `$BOOST_CI_SRC_FOLDER`) as we set that variable in that file and already use it at some places. E.g. we **must** use it for `"$BOOST_CI_SRC_FOLDER/ci/common_install.sh"` and then switching to `$BOOST_ROOT/libs/$SELF/.drone/after-install.sh` would look strange wouldn't it? So from the perspective of those 2 scripts (`drone.*`) it makes sense to keep it and the copy.

sdarwin commented:
>  always use $BOOST_ROOT and pretend $BOOST_CI_SRC_FOLDER doesn't exist anymore

Right, this was what I was thinking, because the tests are being run in $BOOST_ROOT, so to refer back to $BOOST_CI_SRC_FOLDER later on is a bit confusing.   But, if it works either way.

>  moving the library folder instead of copying it

Moving instead of copying could be dangerous, it might have side effects.   Some CI systems expect the main src folder to at least be present.  Copy is safer.
